20_1 DataGrid onRowUpdating: docissues (#1891)
DevExpress/devextreme-documentation: includes/datagrid-ref-rowupdateconfirmandvalidate.md

---

##### jQuery

    <!--JavaScript-->
    $(function(){
        $("#{widgetName}Container").dx{WidgetName}({
            // ...
            onRowUpdating: function(e) {

>                 var deferred = $.Deferred();

Reviewer: ```suggestion
                const deferred = $.Deferred();
```

>                 var promptPromise = DevExpress.ui.dialog.confirm("Are you sure?", "Confirm");

Reviewer: ```suggestion
                const promptPromise = DevExpress.ui.dialog.confirm("Are you sure?", "Confirm changes");
```

                promptPromise.done((dialogResult) => {
                    if (dialogResult) {
                        let params = "";
                        for (let key in e.newData) {
                            params += `${key}=${e.newData[key]}&`;
                        }
                        params = params.slice(0, -1);
                        $.ajax({
                            url: "https://url/to/your/validation/service",
                            dataType: "json",
                            data: params,
                            success: function(validationResult) {

>                             !validationResult.errorText ? deferred.resolve(false) : deferred.reject(validationResult.errorText);

Reviewer: ```suggestion
                                !validationResult.errorText ? deferred.resolve(false) : deferred.reject(validationResult.errorText);
```

                            },
                            error: function() {
                                deferred.reject("Data Loading Error");
                            },
                            timeout: 5000
                        });
                    } else {

>                     deferred.resolve(true);              

Reviewer: ```suggestion
                        deferred.resolve(true);              
```

            e.cancel = new Promise((resolve, reject) => {
                const promptPromise = confirm("Are you sure?", "Confirm changes");
                promptPromise.then((dialogResult) => {
                    if (dialogResult) {
                        let params = new HttpParams();
                        for (let key in e.newData) {
                            params = params.set(key, e.newData[key]);
                        }
                        this.httpClient
                            .get("https://url/to/your/validation/service", { params: params })
                            .toPromise()
                            .then((validationResult) => {
                                !validationResult.errorText ? resolve(false) : reject(validationResult.errorText);
                            })
                            .catch(() => reject());
                    } else {
                        resolve(true);
                    }
                });
            });
        }
    }

    <!-- tab: app.component.html -->
    <dx-{widget-name} ... 
        (onRowUpdating)="updateRow($event)">
    </dx-{widget-name}>

    <!-- tab: app.module.ts -->
    // ... 

>     import { DxDataGridModule } from 'devextreme-angular'; 

Reviewer: ```suggestion
    import { Dx{WidgetName}Module } from 'devextreme-angular'; 
```

    import { HttpClientModule } from "@angular/common/http";

    @NgModule({
        imports: [
            // ...
            Dx{WidgetName}Module,
            HttpClientModule
        ],
        // ...
    })

##### Vue

    <!-- tab: App.vue -->
    <template>

>         <DxDataGrid ...

Reviewer: ```suggestion
        <Dx{WidgetName} ...
```

            @row-updating="updateRow">

>         </DxDataGrid>

Reviewer: ```suggestion
        </Dx{WidgetName}>
```

    </template>
    <script>
    import Dx{WidgetName}, { ... } from 'devextreme-vue/{widget-name}';
    import dialog from 'devextreme/ui/dialog';
    // ...

    export default {
        components: {

>             DxDataGrid,

Reviewer: ```suggestion
            Dx{WidgetName},
```

                    promptPromise.then((dialogResult) => {
                        if (dialogResult) {
                            let params = "?";
                            for (let key in e.newData) {
                                params += `${key}=${e.newData[key]}&`;
                            }
                            params = params.slice(0, -1);
                            fetch(`https://url/to/your/validation/service${params}`)
                                .then((validationResult) => {
                                    !validationResult.errorText ? resolve(false) : reject(validationResult.errorText);
                                })
                                .catch(() => reject());
                        } else {
                            resolve(true);
                        }
                    });
                });
            }
        },
    };
    </script>

##### React

    <!--tab: App.js-->
    import React from 'react';
    import 'devextreme/dist/css/dx.common.css';
    import 'devextreme/dist/css/dx.light.css';
    
    import {WidgetName}, { ... } from 'devextreme-react/{widget-name}';

    function onRowUpdating(e) {

>     e.cancel = new Promise((resolve, reject) => {

Reviewer: The contents of this function are misaligned.

        const promptPromise = dialog.confirm("Are you sure?", "Confirm changes");
        promptPromise.then((dialogResult) => {
        if (dialogResult) {
            let params = "?";
            for (let key in e.newData) {
            params += `${key}=${e.newData[key]}&`;
            }
            params = params.slice(0, -1);
            fetch(`https://url/to/your/validation/service${params}`)
            .then((validationResult) => {
                !validationResult.errorText
                ? resolve(false)
                : reject(validationResult.errorText);
            })
            .catch(() => reject());
        } else {
            resolve(true);
        }
        });
    });
    }

    function App() {

>     return (

Reviewer: Here too.

>         <DataGrid ...

Reviewer: ```suggestion
        <{WidgetName} ...
```

            onRowUpdating={onRowUpdating}>
            // ...

>         </DataGrid>

Reviewer: ```suggestion
        </{WidgetName}>
```
